src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5042,6 +5042,326 @@ class StubGenerator: public StubCodeGenerator {
+    VectorRegister vtable_16 = (MaxVectorSize == 16) ? v27 : v30;
+    VectorRegister vtemp1 = v28; // group: v29, v30, v31
+    VectorRegister vtemp2 = v29;
+
+    // Max number of bytes we can process before having to take the mod
+    // 0x15B0 is 5552 in decimal, the largest n such that 255n(n+1)/2 + (n+1)(BASE-1) <= 2^32-1
+    const uint64_t BASE = 0xfff1;
+    const uint64_t NMAX = 0x15B0;
+
+    __ enter(); // required for proper stackwalking of RuntimeStub frame
+    __ mv(temp1, 64);
+    __ vsetvli(temp0, temp1, Assembler::e8, Assembler::m4);
+
+    // Generating accumulation coefficients for further calculations
+    __ vid_v(vtemp1);
+    __ vmv_v_x(vtable_64, temp1);
+    __ vsub_vv(vtable_64, vtable_64, vtemp1);
+    // vtable_64 group now contains { 0x40, 0x3f, 0x3e, ..., 0x3, 0x2, 0x1 }
+    __ vmv_v_i(vzero, 0);
+    if (MaxVectorSize > 16) {
+      // Need to generate vtable_16 explicitly
+      __ mv(temp1, 16);
+      __ vsetvli(temp0, temp1, Assembler::e8, Assembler::m1);
+
+      // Generating accumulation coefficients for further calculations
+      __ vid_v(vtemp1);
+      __ vmv_v_x(vtable_16, temp1);
+      __ vsub_vv(vtable_16, vtable_16, vtemp1);
+    }
+
+    __ mv(base, BASE);
+    __ mv(nmax, NMAX);
+    __ mv(right_16_bits, right_n_bits(16));
+
+    __ srli(s2, adler, 16); // s2 = ((adler >> 16) & 0xffff)
+    // s1 is initialized to the lower 16 bits of adler
+    // s2 is initialized to the upper 16 bits of adler
+    if (!UseZbb) {
+      __ andr(s2, s2, right_16_bits);
+      __ andr(s1, adler, right_16_bits); // s1 = (adler & 0xffff)
+    } else {
+      __ zext_h(s2, s2);
+      __ zext_h(s1, adler);
+    }
+
+    // The pipelined loop needs at least 16 elements for 1 iteration
+    // It does check this, but it is more effective to skip to the cleanup loop
+    __ mv(temp0, (u1)16);
+    __ bgeu(len, temp0, L_nmax);
+    __ beqz(len, L_combine);
+
+  __ bind(L_simple_by1_loop);
+    __ lbu(temp0, Address(buff, 0));
+    __ addi(buff, buff, 1);
+    __ add(s1, s1, temp0);
+    __ add(s2, s2, s1);
+    __ sub(len, len, 1);
+    __ bgtz(len, L_simple_by1_loop);
+
+    // s1 = s1 % BASE
+    __ remuw(s1, s1, base);
+    // s2 = s2 % BASE
+    __ remuw(s2, s2, base);
+
+    __ j(L_combine);
+
+  __ bind(L_nmax);
+    __ sub(len, len, nmax);
+    __ sub(count, nmax, 16);
+    __ bltz(len, L_by16);
+    __ mv(step, 64);
+
+  // Align L_nmax loop by 64
+  __ bind(L_nmax_loop_entry);
+    __ sub(count, count, 32);
+
+  __ bind(L_nmax_loop);
+    adler32_process_bytes_by64(buff, s1, s2, step, vtable_64, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, vtemp1, vtemp2);
+    __ sub(count, count, 64);
+    __ bgtz(count, L_nmax_loop);
+
+    // There are three iterations left to do
+    const int remainder = 3;
+    adler32_process_bytes_by16(buff, s1, s2, right_16_bits, vtable_16, vzero,
+      vbytes, vs1acc, vs2acc, temp0, temp1, temp2, vtemp1, vtemp2, remainder);
```

Comment:
Maybe deserves another `adler32_process_bytes_by32` here? Then you do one `adler32_process_bytes_by32` and one `adler32_process_bytes_by16` for the rest 3 iterations.

Reply:
Done